### ISSUE TITLE
`Literal` Loses Type when Assigned in Superclass

### ISSUE THREAD

**Issue by adam-grant-hendry** [microsoft/pyright#3651]
**Describe the bug**
I am working with the following source code:

```python
from __future__ import annotations

import typing

Direction = typing.Literal['up', 'down']

class A:
    def set_inst_attrs(self, dir_: Direction = 'up') -> None:
        self.dir_ = dir_
        reveal_type(self.dir_)

class B(A):
    def __init__(self, dir_: Direction = 'down') -> None:
        self.set_inst_attrs(dir_)
        reveal_type(self.dir_)

```

In `mypy`, the `reveal_type` is correct in both cases (`Literal['up', 'down']`). In `pyright`, the type is `str` in class `B`:

[mypy](https://mypy-play.net/?mypy=latest&python=3.8&gist=c6ca70e6977029a55b0aa66b6b4eb5ab)

[pyright](https://pyright-playground.decorator-factory.su/?gzip=H4sIAOs1wmIC_4WP3QrCMAyF7_sUuWsHcw8wUJh4Kb6ASChbJoXZjjZTfHtbN-cPirkJSU5OvrTenQCxHXjwhAjm1DvPoK11rNk4G4SYenztjT0KsTGe6jSC5dQrtobJ624vh17mIBt3sfIgRN3pEKAqBcRoqIVAjMYGRs3sgwrUtTk0xmMJr67JJoPFCnbO0ridIsmLpI6SlOaBpzPpDiMMqVmUPe6vVZU9ETACGEZUb7b5XH2hub8zKn5RfTw23v-HdwNRQ8YgfAEAAA%3D%3D)

The same occurs in VSCode as well when using `pyright`. I'm unsure if this is a difference between the linters or a genuine error in `pyright`, but I would expect the types to remain the same.

**To Reproduce**
**OS:** `Windows 10, x64-bit`
**Python:** `3.8.10 x64-bit`
**VSCode:** `1.68.1`
**mypy:** `0.961`
**pyright:** `1.1.256`

**Expected behavior**
`Literal` types do not lose their type when assigned in superclasses (see code example for context).

**Screenshots or Code**
### Mypy Playground
![image](https://user-images.githubusercontent.com/59346180/177063679-73402d96-bcf1-43f9-b882-3bfc76b26516.png)

### Pyright Playground
![image](https://user-images.githubusercontent.com/59346180/177063756-701217d6-42d5-4343-8512-33475ed6691c.png)

### VSCode
![image](https://user-images.githubusercontent.com/59346180/177063890-75796bcb-56a1-4b51-9767-5812e3677f8d.png)

**VS Code extension or command-line**
Pylance v2022.6.30

**Comment by erictraut** [comment]
Pyright is working as intended here.

You have not declared the type for the instance variable `dir_`, so pyright infers its type based on normal inference rules. In this case, its inferred type is `str`. Locally, within `A.set_inst_attrs`, the type is further narrowed based on the assignment, which is why the `reveal_type` immediately after the assignment reveals a narrower type.

You can see the same thing if you assign a literal string expression.

```python
class A:
    def __init__(self):
        self.foo = "up"
        reveal_type(self.foo) # Literal["up"]

reveal_type(A().foo) # str
```

If your intent is for `dir_` to be constrained to the type `Direction`, you need to provide a type declaration.

```python
    def set_inst_attrs(self, dir_: Direction = "up") -> None:
        self.dir_: Direction = dir_
        reveal_type(self.dir_)
```

**Comment by adam-grant-hendry** [comment]
@erictraut Your example is wrong.

> You have not declared the type for the instance variable `dir_`, so pyright infers its type based on normal inference rules.

That's not true. The types are declared in the function signatures. Types are often inferred from function signatures.

> In this case, its inferred type is str

That is also not true. The type of the agrument `dir_` passed to `self.dir_` is revealed to be `Literal['up', 'down']. See below:

![image](https://user-images.githubusercontent.com/59346180/177190935-a0402c60-fde9-422b-bfef-fd294744ea30.png)

**Comment by erictraut** [comment]
> The types are declared in the function signatures

The type of parameter `dir_` is declared in the function signature. That's a local variable within the `set_inst_attrs` scope. It's a different symbol than the `dir_` instance variable. You have not provided a type declaration for the other. The fact that they happen to share the same name is throwing you off. Consider if you had named the input parameter `x` instead of `dir_`, and that might make things clearer for you.

```python
class A:
    def set_inst_attrs(self, x: Direction = 'up') -> None:
        self.dir_ = x
```

> The type of the agrument dir_ passed to self.dir_ is revealed to be `Literal['up', 'down'].

The `reveal_type` function reveals the type of an expression _at that point in the code flow`. That can be different (narrower) than the type of a particular symbol. Consider the following:

```python
def func(x: str | int, y: str):
    reveal_type(x) # str | int
    if isinstance(x, int):
        reveal_type(x) # int
    else:
        reveal_type(x) # str
    reveal_type(x) # int | str
    x = y
    reveal_type(x) # str
    x = "hi"
    reveal_type(x) # Literal["hi"]
```

Note that the revealed type changes throughout the function due to narrowing. The type of variable `x` is declared to be `str | int`, but at any point in the code flow, the type checker may be able to determine that it contains a narrower type.

In your example, the type of instance variable `dir_` has no type declaration, so its type is inferred based on assignments to it. In this case, its type is inferred to be `str`. You wouldn't generally want a variable to be inferred as a literal because that's too constraining. Consider the following:

```python
class A:
    def __init__(self):
        self.foo = 1

a = A()
a.foo = 2 # This would generate an error if the type of `foo` was inferred to be `Literal[1]`
```

So, going back to your example, the inferred type of instance variable `dir_` is `str`. Locally, its type is further narrowed based on the assignment, but in other execution scopes, its type is `str` unless/until it is further narrowed within that scope.

This is all working as designed. It is not a bug.

For additional details, please refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md).

**Comment by adam-grant-hendry** [comment]
@erictraut Isn't a `Literal` more narrow than `str`? Furthermore, `Literal`s can accept different types:

```python
yes: Literal[True, 1, 'yes'] = True
```

This from the docs doesn't make sense. `x` should be an `int`, not a `Literal`. It hasn't been declared as a `Literal`:

```python
x = 1
reveal_type(x)  # Type of "x" is "Literal[1]"
```

**Comment by erictraut** [comment]
We're getting a bit off topic for this issue. I recommend creating a discussion thread if you have additional questions about how static typing works.

Quick answers to your above questions...

> Isn't a Literal more narrow than `str`

A string literal is narrower than `str`.

> Literals can accept different types

The `Literal` special form accepts different literal expressions. This is translated to a union of literal types (`Literal[True] | Literal[1] | Literal['yes']`.

> `x` should be an `int`, not a `Literal`

The inferred type of `x` is `int`, but the revealed type is a literal at that point in the code flow because the type has been narrowed on assignment.

**Comment by adam-grant-hendry** [comment]
> about how static typing works.

I think you should clarify that this is how static typing works _with the pyright static type checker_. The behavior is different with the _mypy static type checker_.

You are implying things are general and `pyright` takes a different approach from `mypy`, much like Facebook's `pyre`, Google's `pytype`, and what's built into JetBrain's `PyCharm` IDE.

`pyright` is opinionated, and what works for it might not work for other checkers.

"How static typing works" is directly related to the static type checker used and its interpretation of `PEP 483` and `PEP 484`.

**Comment by erictraut** [comment]
You are posting questions in the pyright issue tracker and asking questions about the pyright documentation, so I presume we are talking about static typing in the context of pyright. If your questions are more generally about the typing standards that all of the Python type checkers are based on (PEP 484, etc.), then the [python/typing discussion forum](https://github.com/python/typing/discussions) would be a more appropriate place to post your questions.
